### ISSUE TITLE
Figure out Performance Bottlenecks for stack analyses

### ISSUE THREAD

**Issue by samuzzal-choudhury** [openshiftio/openshift.io#1647]
#### User Story
As an OSIO developer, I will be able to get stack analysis results within the accepted SLA so that I can refactor my stack.

#### Description
Since the current stack analyses doesn't respect the SLA, we would like to figure out the performance bottleneck(s) in the stack analyses workflow so that we can tune the identified performance bottleneck(s).

#### Acceptance Criteria
- All the performance bottleneck(s) for stack analyses are identified.
- All the performance tuning areas for identified bottleneck(s) are identified.

#### Primary Task List (2 sets of tests by Sam and Aagam)
- [x] Load test for 100 concurrent users by avoiding PGM call from RecommenderTask 
- [ ] Load test for 100 concurrent users by hardcoding output of GraphAggregatorTask
- [x] Load test for 100 concurrent users by avoiding License server call in extract_user_stack_package_licenses from RecommenderTask
- [x] Load test for 100 concurrent users by avoiding companion package version retrieval from RecommenderTask
- [x] Load test for 100 concurrent users by avoiding alternate package version retrieval from RecommenderTask
- [x] Load test for 100 concurrent users by avoiding License server call in StackAggregatorTask

#### Follow up tasks
- [x] Identify the performance bottleneck(s)
- [x] Identify the performance tuning areas for bottleneck(s)
- [x] Fix service(s) causing bottleneck (Stretch goal)

**Comment by pkajaba** [comment]
@samuzzal-choudhury why don't we do identifying the performance bottlenecks in parallel with testing? I am following a discussion about stack analysis and scaling of it very closely. I believe that scaling is really important to scale containers. It will improve some performance for sure. However, I believe that main issues/bottleneck is in PGM/Kronos implementation.  

From what I remember it was rushed and even though there were some changes core of Kronos remained unchanged.

**Comment by samuzzal-choudhury** [comment]
@pkajaba That's what we are planning to do. Here is the overall plan of actions https://docs.google.com/spreadsheets/d/1NALaE5uMbflPd_Q7-Eto-OT43NovEOrY83qV5xvRyR0/edit?ts=5a320619#gid=0. Hope this helps.

**Comment by samuzzal-choudhury** [comment]
A bug was found during the tests https://github.com/openshiftio/openshift.io/issues/1704.

**Comment by samuzzal-choudhury** [comment]
The result for the load test for 100 concurrent users by avoiding PGM call from RecommenderTask: 
![stack_analysis_parallel_calls_100_threads_concurrent_100_threads_min_max_avg_times](https://user-images.githubusercontent.com/17740190/34148797-c36f66f2-e4c8-11e7-9358-fa4628e389ec.png)
![stack_analysis_parallel_calls_100_threads](https://user-images.githubusercontent.com/17740190/34148798-c39f9412-e4c8-11e7-8b0c-351ebb1a8d20.png)
![stack_analysis_parallel_calls](https://user-images.githubusercontent.com/17740190/34148799-c3d04c56-e4c8-11e7-8ab5-c7a06ad141d4.png)

**Comment by samuzzal-choudhury** [comment]
The result of the load test for 100 concurrent users by hardcoding output of GraphAggregatorTask

![stack_analysis_parallel_calls_100_threads_concurrent_100_threads_min_max_avg_times](https://user-images.githubusercontent.com/17740190/34149971-f1a8dfe0-e4cc-11e7-84d3-73570116cda7.png)
![stack_analysis_parallel_calls_100_threads](https://user-images.githubusercontent.com/17740190/34149974-f2391a88-e4cc-11e7-8c03-4bf5b3c319d3.png)
![stack_analysis_parallel_calls](https://user-images.githubusercontent.com/17740190/34149975-f2f48520-e4cc-11e7-83a1-4b353fb1b8fe.png)

**Comment by samuzzal-choudhury** [comment]
The result of the load test for 100 concurrent users by hardcoding output of GraphAggregatorTask with 16 worker-api pods.

![stack_analysis_parallel_calls_100_threads_concurrent_100_threads_min_max_avg_times](https://user-images.githubusercontent.com/17740190/34150251-d9f7027c-e4cd-11e7-8f7d-b8e7c95295ab.png)
![stack_analysis_parallel_calls_100_threads](https://user-images.githubusercontent.com/17740190/34150253-da291b36-e4cd-11e7-9505-9a1ed5ffe50a.png)
![stack_analysis_parallel_calls](https://user-images.githubusercontent.com/17740190/34150254-da5a822a-e4cd-11e7-9a1a-55d8badb9c61.png)

**Comment by samuzzal-choudhury** [comment]
License server response time was consistently between 6-10 milliseconds.

Some examples -
```
------------------------------------------------------
Request ID:  4939f2446fc8487d9200c1e8b6a5a2b6
------------------------------------------------------
Stack Aggregator Graph Call time taken(ms): 226.85009765625
Stack Aggregator License Call time taken(ms): 6.385009765625
Recommender PGM Call time taken(ms): 0.114990234375
Recommender First Graph Call time taken(ms): 598.89697265625
Recommender Second Graph Call time taken(ms): 62.77392578125
Recommender Third Graph Call time taken(ms): 82.68798828125
Recommender License Call time taken(ms): 6.30810546875
Total time taken(ms): 1016.695068359375
------------------------------------------------------
------------------------------------------------------
Request ID:  2b4e3324b17043a3aac604d338eee4c5
------------------------------------------------------
Stack Aggregator Graph Call time taken(ms): 82.948974609375
Stack Aggregator License Call time taken(ms): 8.84912109375
Recommender PGM Call time taken(ms): 0.179931640625
Recommender First Graph Call time taken(ms): 447.944091796875
Recommender Second Graph Call time taken(ms): 27.01513671875
Recommender Third Graph Call time taken(ms): 79.469970703125
Recommender License Call time taken(ms): 6.117919921875
Total time taken(ms): 656.908935546875
------------------------------------------------------
------------------------------------------------------
Request ID:  23b089f6f52d4fcbbc9c3c6bed78f114
------------------------------------------------------
Stack Aggregator Graph Call time taken(ms): 103.14794921875
Stack Aggregator License Call time taken(ms): 6.975830078125
Recommender PGM Call time taken(ms): 0.091064453125
Recommender First Graph Call time taken(ms): 452.62890625
Recommender Second Graph Call time taken(ms): 26.66796875
Recommender Third Graph Call time taken(ms): 85.310791015625
Recommender License Call time taken(ms): 5.89697265625
Total time taken(ms): 684.516845703125
------------------------------------------------------

```

**Comment by abs51295** [comment]
Outlier detection was identified as a performance bottleneck for PGM and a new logic was set in place. The documentation for the same can be found at: https://docs.google.com/a/redhat.com/document/d/1kLyEbJDY8LuCU5nr2kvcHivCFm-bAEe-38bI_IkeCfs/edit?usp=sharing. Also, the code can be found at: https://github.com/abs51295/fabric8-analytics-stack-analysis/tree/outlier_detection

**Comment by pkajaba** [comment]
> documentation for the same can be found ...

Why can't you add to the repository?

>  Also, the code can be found ...

Why don't you create PR to have working code without bottlenecks inside the main repository? (I am sorry if you did, but I could not find it)

**Comment by abs51295** [comment]
@pkajaba The PR has already been created and merged: https://github.com/fabric8-analytics/fabric8-analytics-stack-analysis/pull/168. Also, it has been pushed to prod: https://github.com/openshiftio/saas-analytics/pull/193. Regarding the doc, since the approach was still under discussion I didn't add it to the repo, but since the PR has been merged now, I would add it to the repo. Thanks.

**Comment by pkajaba** [comment]
sure

**Comment by samuzzal-choudhury** [comment]
The performance test results can be found here: https://docs.google.com/spreadsheets/d/160SEJNO__1UdCTW62s0x18YjsepApgosvlbVbwt-kSs/edit#gid=1208181716

Closing this issue.
